contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+   }
+
+   void system_contract::regfinkey( const name& producer, const std::string& finalizer_key, const std::string& proof_of_possession) {
+      require_auth( producer );
+
+      check( is_registered_producer(producer), "The producer is not a registered producer");
+
+      // Basic key and signature format check
+      check(finalizer_key.compare(0, 7, "PUB_BLS") == 0, "finalizer key must start with PUB_BLS");
+      check(proof_of_possession.compare(0, 7, "SIG_BLS") == 0, "proof of possession siganture must start with SIG_BLS");
+
+      // Proof of possession check
+      const auto pk = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      const auto signature = eosio::decode_bls_signature_to_g2(proof_of_possession);
+      check(eosio::bls_pop_verify(pk, signature), "proof of possession check failed");
+
+      // Todo: Go over finalizer_keys_table to make sure no duplicate keys are
+      // registered.
+
+      auto it = _finalizer_keys.find( producer.value );
+      if( it != _finalizer_keys.end() ) {
+         _finalizer_keys.modify( it, same_payer, [&]( auto& k ) {
+            k.registered_finalizer_keys.insert(finalizer_key);
+         });
+      } else {
+         // The first time the producer registers a finalizer key
+         _finalizer_keys.emplace( producer, [&]( auto& p ) {
+            k.producer = producer;
+            k.registered_finalizer_keys.insert(finalizer_key);
+            k.active_key = finalizer_key;
+         });
+
+         if( is_active_producer(producer) ) {
+            replace_key_in_finalizer_policy(producer, finalizer_key);
+         }
+      }
+   }
+
+   void system_contract::actfinkey( const name& producer, const std::string& finalizer_key ) {
+      require_auth( producer );
+
+      check( is_registered_producer(producer), "The producer is not a registered producer");
+
+      auto it = _finalizer_keys.find( producer.value );
+      check(it != _finalizer_keys.end(), "producer has not registered any keys");
+      
+      check(it->registered_finalizer_keys.find(finalizer_key) != it->registered_finalizer_keys.end(), "finalizer_key was not registered");
+      
+      _finalizer_keys.modify( it, same_payer, [&]( auto& k ) {
+         k.active_finalizer_key = finalizer_key;
+      });
+
+      if( is_active_producer(producer) ) {
+         replace_key_in_finalizer_policy(producer, finalizer_key);
+      }
+   }
+
+   void system_contract::delfinkey( const name& producer, const std::string& finalizer_key ) {
+      require_auth( producer );
+
+      check( is_registered_producer(producer), "The producer is not a registered producer");
+
+      auto it = _finalizer_keys.find( producer.value );
+      check(it != _finalizer_keys.end(), "producer has not registered any keys");
+      
+      check(it->registered_finalizer_keys.find(finalizer_key) != it->registered_finalizer_keys.end(), "finalizer_key was not registered");
+      
+      _finalizer_keys.erase( it );
```

Comment:
We need to disallow erasing an active finalizer key unless it is the last registered finalizer key of the BP.